ggml/src/ggml-vulkan/ggml-vulkan.cpp

```diff
@@ -7433,9 +7433,13 @@ static void ggml_vk_cleanup(ggml_backend_vk_context * ctx) {
 }
 
 static int ggml_vk_get_device_count() {
-    ggml_vk_instance_init();
-
-    return vk_instance.device_indices.size();
+    try {
+        ggml_vk_instance_init();
+        return vk_instance.device_indices.size();
+    } catch (const vk::SystemError& e) {
+        std::cerr << "ggml_vulkan: Error: System error " << e.what() << std::endl;
```

Comment:
Would be good for this to explicitly say something like "will fallback to CPU".

Reply:
Would be a good idea, but I don't know as the backend what the user of 4 different backends at the same time willl do.

Reply:
We could adapt ggml/src/ggml-backend-reg.cpp

```
    void register_backend(ggml_backend_reg_t reg, dl_handle_ptr handle = nullptr) {
        if (!reg) {
            return;
        }

#ifndef NDEBUG
        GGML_LOG_DEBUG("%s: registered backend %s (%zu devices)\n",
            __func__, ggml_backend_reg_name(reg), ggml_backend_reg_dev_count(reg));
#endif
        backends.push_back({ reg, std::move(handle) });
        for (size_t i = 0; i < ggml_backend_reg_dev_count(reg); i++) {
            register_device(ggml_backend_reg_dev_get(reg, i));
        }
    }
```

to interpret ggml_backend_reg_dev_count(reg) returning 0 as "oops, don't use me". Eventually, as we have registered no device at all even though we tried we could say we are now using CPU only.

Reply:
Some backends intentionally may have zero devices, for example the RPC backend does not have a device list by itself, they need to be created by the user. However returning `NULL` for backends where this is not possible can be more efficient, since it will cause the backend to be unloaded completely when using `GGML_BACKEND_DL`. So that would be the preferred option.

Reply:
For what it's worth, when I enable GGML_BACKEND_DL, the vulkan backend file disappears entirely from the installation.

Reply:
Reading through ggml-vulkan.cpp, it seems the intention is to late bind which vulkan instance exactly (as long as possible). There's a mysterious comment

```
// Should be changed to return device-specific host buffer type
// but that probably requires changes in llama.cpp
```

Reply:
> I think libggml-vulkan.so moves from lib to bin and cmake install doesn't know that or something.

When `GGML_BACKEND_DL` is enabled, backends are built as `MODULE` targets instead of library, and one of the consequences is they go into the `RUNTIME` directory instead. It's not very clear where they should be installed, currently ggml only looks for backends in the same directory as the executable, so for it to even work, they would need to be installed in the `bin` directory, which is not great. So at the moment this is only useful for applications that handle backend loading themselves, but not as installable libraries.

Reply:
That comment is just about host buffers, not immediately relevant to this.

The intention was not to defer the decision, but at the time of writing it was unclear which function would get called first, so there's a number of options that trigger initializing the instance. Not sure if that has changed.

I think it's a good idea to leave a comment about not having found any Vulkan devices or failing to initialize the instance, but you should probably use the GGML debug macro for that instead of piping to `std::cerr`.